test_requests.py

```diff
@@ -1265,6 +1265,25 @@ def test_get_environ_proxies(self):
             'http://localhost.localdomain:5000/v1.0/') == {}
         assert get_environ_proxies('http://www.requests.com/') != {}
 
+    def test_guess_filename_when_int(self):
+        from requests.utils import guess_filename
+        assert None is guess_filename(1)
+
+    def test_guess_filename_when_filename_is_an_int(self):
+        from requests.utils import guess_filename
+        fake = type('Fake', (object,), {'name': 1})()
+        assert None is guess_filename(fake)
+
+    def test_guess_filename_with_file_like_obj(self):
+        from requests.utils import guess_filename
+        fake = type('Fake', (object,), {'name': 'value'})()
+        assert 'value' == guess_filename(fake)
+
+    def test_guess_filename_with_unicode_name(self):
+        from requests.utils import guess_filename
+        fake = type('Fake', (object,), {'name': b'value'.decode('utf-8')})()
+        assert 'value' == guess_filename(fake)
```

Comment:
The test will work fine on Py2, it just won't prove anything because of the implicit string type conversion. The test fundamentally asks a different question on each platform. On Python 3, it will say "does guess_filename return a unicode string with the value 'value'", whereas on Python 2 it will say "does guess_filename return any string that can be coerced to 'value'".

Observe that if this was the function of `guess_filename` it would work fine on Py 2 and fail on Py 3:

```python
def guess_filename(obj):
    return b'value'
```

Note that the check against `None` is included in a type check. I'm happy to check the value _as well_, but we need to write a test that Python 2's implicit string conversion doesn't defeat.